flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecJoinBase.scala

```diff
@@ -70,4 +76,150 @@ abstract class BatchExecJoinBase(
+    *         The third element is the distribution requirement of right child if the requirement
+    *         can be push down into join.
+    */
+  def satisfyHashDistributionOnNonBroadcastJoin(
+      requiredDistribution: FlinkRelDistribution)
+  : (Boolean, FlinkRelDistribution, FlinkRelDistribution) = {
+    // Only Non-broadcast HashJoin could provide HashDistribution
+    if (requiredDistribution.getType != HASH_DISTRIBUTED) {
+      return (false, null, null)
+    }
+    // Full outer join cannot provide Hash distribute because it will generate null for left/right
+    // side if there is no match row.
+    if (joinType == JoinRelType.FULL) {
+      return (false, null, null)
+    }
+
+    val leftKeys = joinInfo.leftKeys
+    val rightKeys = joinInfo.rightKeys
+    val leftKeysToRightKeys = leftKeys.zip(rightKeys).toMap
+    val rightKeysToLeftKeys = rightKeys.zip(leftKeys).toMap
+    val leftFieldCnt = getLeft.getRowType.getFieldCount
+    val requiredShuffleKeys = requiredDistribution.getKeys
+    val requiredLeftShuffleKeys = mutable.ArrayBuffer[Int]()
+    val requiredRightShuffleKeys = mutable.ArrayBuffer[Int]()
+    requiredShuffleKeys.foreach { key =>
+      if (key < leftFieldCnt && joinType != JoinRelType.RIGHT) {
+        leftKeysToRightKeys.get(key) match {
+          case Some(rk) =>
+            requiredLeftShuffleKeys += key
+            requiredRightShuffleKeys += rk
+          case None if requiredDistribution.requireStrict =>
+            // Cannot satisfy required hash distribution due to required distribution is restrict
+            // however the key is not found in right
+            return (false, null, null)
+          case _ => // do nothing
+        }
+      } else if (key >= leftFieldCnt &&
+        (joinType == JoinRelType.RIGHT || joinType == JoinRelType.INNER)) {
+        val keysOnRightChild = key - leftFieldCnt
+        rightKeysToLeftKeys.get(keysOnRightChild) match {
+          case Some(lk) =>
+            requiredLeftShuffleKeys += lk
+            requiredRightShuffleKeys += keysOnRightChild
+          case None if requiredDistribution.requireStrict =>
+            // Cannot satisfy required hash distribution due to required distribution is restrict
+            // however the key is not found in left
+            return (false, null, null)
+          case _ => // do nothing
+        }
+      } else {
+        // cannot satisfy required hash distribution if requirement shuffle keys are not come from
+        // left side when Join is LOJ or are not come from right side when Join is ROJ.
+        return (false, null, null)
+      }
+    }
+    if (requiredLeftShuffleKeys.isEmpty) {
+      // the join can not satisfy the required hash distribution
+      // due to the required input shuffle keys are empty
+      return (false, null, null)
+    }
+
+    val (leftShuffleKeys, rightShuffleKeys) = if (joinType == JoinRelType.INNER &&
+      !requiredDistribution.requireStrict) {
```

Comment:
format, you should set `Continuation indent` for scala to 4

Reply:
ok